### PR TITLE
Check how long until radical guru (Fixes #222)

### LINKED ISSUE #222: Funky time remaining estimate
There's still something weird with the level up time, since for this level I actually need 30 kanji done, which is not possible in the amount of time it display. Screenshot below.
signal-2019-12-01-182530

### PR DESCRIPTION
**Pull request by turtlemaster19**
To attempt to fix #222, this code attempts to check the last radical guru time if a kanji is still locked.

**EDIT:** Thanks for helping with CircleCI! I was trying to work on this issue on vacation, so I couldn't use my computer to test the builds.

### CODE REVIEW

**Comment by turtlemaster19** [comment]
I am going to mark it ready for review. Please check that this works, it’s failing CircleCI because there’s a fastlane update?!

**Comment by davidsansome** [comment]
You've gotta scroll up in that test output box in CircleCI, or click the icon on the right to open the log in a new window.  Here's the full output: https://app.circleci.com/jobs/github/davidsansome/tsurukame/39/parallel-runs/0/steps/0-106

**Comment by turtlemaster19** [comment]
Oops! Thank you!

**Comment by turtlemaster19** [comment]
@davidsansome Ready for review now. Thanks for helping with CircleCI!

File: ios/LevelTimeRemainingItem.swift [hunk 1]
```diff
@@ -50,40 +50,60 @@ class LevelTimeRemainingCell: TKMModelCell {
...
+    for assignment in item.currentLevelAssignments {
+      if assignment.subjectType != .radical {
+        continue
+      }
+      guard let subject = item.services.dataLoader.load(subjectID: Int(assignment.subjectId)) else {
+        continue
+      }
+      radicalDates.append(assignment.guruDate(for: subject))
+    }
+    var lastRadicalGuruTime = radicalDates.last!.timeIntervalSinceNow
```
Comment (davidsansome) [inline ios/LevelTimeRemainingItem.swift]: The `last!` here will crash if the array is empty for whatever reason.
What about:

    radicalDates.last?.timeIntervalSinceNow ?? 0

File: ios/LevelTimeRemainingItem.swift [hunk 2]
```diff
@@ -50,40 +50,60 @@ class LevelTimeRemainingCell: TKMModelCell {
...
+      levels.append(assignment.level)
     }
 
     // Sort the list of dates and remove the most distant 10%.
     guruDates.sort()
     guruDates.removeLast(Int(Double(guruDates.count) * 0.1))
+    levels = Array(levels.sorted().reversed())
+    levels.removeLast(Int(Double(levels.count) * 0.1))
 
     if let lastDate = guruDates.last {
```
Comment (davidsansome) [inline ios/LevelTimeRemainingItem.swift]: Add `levels.last!` to the guard here as well so it doesn't crash if levels is empty:

    if let lastData = guruDates.last, lastLevel = levels.last {
    ...
      average = max(average, TKMMinimumTimeUntilGuruSeconds(lastLevel, 1) + lastRadicalGuruTime)

File: ios/LevelTimeRemainingItem.swift [hunk 3]
```diff
@@ -50,40 +50,60 @@ class LevelTimeRemainingCell: TKMModelCell {
...
+        // they've spent at this level so far, as an estimate.
+        var average = item.services.localCachingClient!.getAverageRemainingLevelTime()
+
+        // But ensure it can't be less than the time it would take to get a fresh item
+        // to Guru, if they've spent longer at the current level than the average.
+        average = max(average, TKMMinimumTimeUntilGuruSeconds(levels.last!, 1) + lastRadicalGuruTime)
+
+        setRemaining(Date(timeIntervalSinceNow: average), isEstimate: true)
+      }
+      else {
```
Comment (davidsansome) [inline ios/LevelTimeRemainingItem.swift]: The else should go on the line above.  XCode should automatically format the files for you when you save.

File: ios/LevelTimeRemainingItem.swift (hunk 1 above)
Comment (turtlemaster19) [reply ios/LevelTimeRemainingItem.swift]: Fixed it!

File: ios/LevelTimeRemainingItem.swift (hunk 2 above)
Comment (turtlemaster19) [reply ios/LevelTimeRemainingItem.swift]: Fixed it! Apparently there has to be another let:
`let lastGuruDate = guruDates.last, let lastKanjiLevel = levels.last`

File: ios/LevelTimeRemainingItem.swift (hunk 3 above)
Comment (turtlemaster19) [reply ios/LevelTimeRemainingItem.swift]: Done! Sorry, I am going to be away from my computer until Monday.

**Comment by turtlemaster19** [comment]
@davidsansome As I stated in my reviews, I won’t be at my computer until Monday, but I have taken care of the issues you have addressed.

**Comment by turtlemaster19** [comment]
@davidsansome I was able to build locally. It works better after a simple change! Ready to merge.

**Comment by turtlemaster19** [comment]
@davidsansome I was able to build locally. It works better after a simple change! Ready to merge.

File: ios/LevelTimeRemainingItem.swift [hunk 4]
```diff
@@ -50,40 +50,54 @@ class LevelTimeRemainingCell: TKMModelCell {
...
     // Sort the list of dates and remove the most distant 10%.
-    guruDates.sort()
-    guruDates.removeLast(Int(Double(guruDates.count) * 0.1))
-
-    if let lastDate = guruDates.last {
-      setRemaining(lastDate, isEstimate: false)
+    guruDates = Array(guruDates.sorted().dropLast(Int(Double(guruDates.count) * 0.1)))
+    levels = Array(Array(levels.sorted().reversed()).dropLast(Int(Double(levels.count) * 0.1)))
+
+    if let lastGuruDate = guruDates.last, let WKLevel = levels.last {
```
Comment (davidsansome) [inline ios/LevelTimeRemainingItem.swift]: Local variables should start with a lower-case letter, so WKLevel should be wkLevel

File: ios/LevelTimeRemainingItem.swift [hunk 5]
```diff
@@ -50,40 +50,54 @@ class LevelTimeRemainingCell: TKMModelCell {
...
-      if assignment.subjectType != .kanji {
+      if assignment.subjectType != .radical {
+        continue
+      }
+      guard let subject = item.services.dataLoader.load(subjectID: Int(assignment.subjectId)) else {
         continue
       }
+      radicalDates.append(assignment.guruDate(for: subject))
+    }
+    let lastRadicalGuruTime = radicalDates.last?.timeIntervalSinceNow ?? 0
```
[the remark this answers is not in the document]
Comment (turtlemaster19) [reply ios/LevelTimeRemainingItem.swift]: Of course! Why did I forget that? Fixed it!

File: ios/LevelTimeRemainingItem.swift (hunk 4 above)
Comment (turtlemaster19) [reply ios/LevelTimeRemainingItem.swift]: Sorry, I'm not familiar with Swift conventions! Fixed it!

**Comment by turtlemaster19** [comment]
@davidsansome Ready to merge.

**Pull request merged**